plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -898,13 +946,18 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 			self->fw_key_addr[1] = 0xE000;
 			self->code_size[0] = self->code_size[1] = self->fw_data_total_count;
 		}
+		self->dynamic_ts.bonding =
+		    _fu_genesys_revert_hex_char_to_int(self->dynamic_ts.bonding);
+		if (self->static_ts.tool_string_version < TOOL_STRING_VERSION_BONDING_QC)
```

Comment:
```
ts_version = g_ascii_xdigit_value(self->static_ts.tool_string_version);
if (ts_version < ...)
```

Reply:
I will change the definition in enum FuGenesysToolStringVersion to fulfill the usage with the origin value.

Reply:
Please check [ac1f96a](https://github.com/fwupd/fwupd/pull/4426/commits/ac1f96aedb70f22369f90edadb7caf8b45c6f07e).

Reply:
Well, I am confused about the structure `FuGenesysToolStringVersion`.

``` c
typedef enum {
	TOOL_STRING_VERSION_9BYTE_DYNAMIC,
	TOOL_STRING_VERSION_BONDING,
	TOOL_STRING_VERSION_BONDING_QC,
	TOOL_STRING_VERSION_VENDOR_SUPPORT,
	TOOL_STRING_VERSION_MULTI_TOKEN,
	TOOL_STRING_VERSION_2ND_DYNAMIC,
	TOOL_STRING_VERSION_RESERVED,
	TOOL_STRING_VERSION_13BYTE_DYNAMIC,
 } FuGenesysToolStringVersion;
 ```
 
The following change would break the line 808 `if (self->static_ts.tool_string_version >= TOOL_STRING_VERSION_VENDOR_SUPPORT) {`

 ``` diff
-	TOOL_STRING_VERSION_9BYTE_DYNAMIC,
+	TOOL_STRING_VERSION_9BYTE_DYNAMIC = 0x30, /* '0' */
```

All the enum values are shifted by `0x30`.

You may want to set all correct values or move them as `#define`.

Reply:
The values are enumerated as the same as the string descriptor reported.
If I want  a series of definition, should I use `#define` or `enum`?

Reply:
Could you list all the values and there "names"?

Reply:
> Could you list all the values and there "names"?

The `FuGenesysToolStringVersion` is enough for current usage.
In-house firmware spec, they are separated in different files. The  `FuGenesysToolStringVersion` is re-translated by my opinion. I have to search and collect again to update them.

Do you have `GL Tool String v1.XX.pdf` on hand?

Reply:
> Do you have GL Tool String v1.XX.pdf on hand?

Sorry, I don't; I have just understood why you shift if by 0x30, which is ascii for '0' as suggested by your comment. Maybe we should make the comment more explicit.

Reply:
> Maybe we should make the comment more explicit.

Current comment is `/* start at 0x30('0'), align to replied values. */`
How about change to `/* start at 0x30('0'), align to tool string replied values. */`?

Reply:
OOI, '0' is for `0`; after the nineth version, '9' what will be the next TS value? 'A` or 'a'?

The enum would be broken. 0x30 -> 0x41 or 0x61.

Reply:
The values defined are not over '9'. I will discuss your question in-house.

Reply:
> The values defined are not over '9'. I will discuss your question in-house.

At least not yet :D

Reply:
I would prefer to stick to the original enum (without the `= 0x30`) and use `_tsdigit()` when necessary.

Reply:
OK, I will do that.